File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String st) {
        Stack<Character> s = new Stack<>();
        
        for(int i = 0; i<st.length(); i++){
            char ch = st.charAt(i);
            if(ch == '(' || ch == '[' || ch == '{'){
                s.push(ch);
            }
            else{
                if(!s.empty()){
                    char top = s.peek();
                    if((ch == ')' && top == '(') || (ch == ']' && top == '[') || (ch == '}' && top == '{')){
                        s.pop();
                    }
                    else
                        return false;
                }
                else
                    return false;
            } 
        }
        if(s.empty())
            return true;
        else
            return false;
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        for(auto &ch: s){
            if(ch == '(' || ch == '[' || ch == '{')
                st.push(ch);
            else if(ch == ')'){
                if(st.empty() || st.top() != '(')
                    return false;
                st.pop();
            }
            else if(ch == ']'){
                if(st.empty() || st.top() != '[')
                    return false;
                st.pop();
            }
            else{
                if(st.empty() || st.top() != '{')
                    return false;
                st.pop();
            }
        }
        if(!st.empty())
            return false;
        return true;
    }
};